File: src/main/java/frc/utils/rumble/RumbleHandler.java
package frc.utils.rumble;

import java.util.ArrayList;
import java.util.Collection;

import org.littletonrobotics.junction.Logger;

import edu.wpi.first.wpilibj.XboxController;
import edu.wpi.first.wpilibj.GenericHID.RumbleType;

public class RumbleHandler {
    private XboxController controller;
    private double port;

    private boolean following = false;
    private RumbleHandler lead;

    private ArrayList<Rumble> queue = new ArrayList<Rumble>();
    public RumbleHandler(XboxController controller){
        this.controller = controller;
        this.port = controller.getPort();
    }
    public RumbleHandler(RumbleHandler lead){
        this.following = true;
        this.lead = lead;
    }
    /**
     * clear rumble queue, effectivly stopping all rumble
     */
    public void clearQue(){
        queue.clear();
    }
    public void addToQue(Rumble[] a){
        for(Rumble b : a){
            queue.add(b);
        }
    }
    public void addToQue(Collection<Rumble> a){
        for(Rumble b : a){
            queue.add(b);
        }
    }
    public void addToQue(Rumble a){
        queue.add(a);
    }

    public void overrideQue(Rumble a){
        clearQue();
        queue.add(a);
    }
    public void overrideQue(Collection<Rumble> a){
        clearQue();
        for(Rumble b : a){
            queue.add(b);
        }
    }
    public void overrideQue(Rumble[] a){
        clearQue();
        for(Rumble b : a){
            queue.add(b);
        }
    }
    public void update(double loopTime){
        if(following){
            queue = lead.queue;
        }
        if(queue.size() > 0){
            queue.get(0).time -= loopTime;
        }
        for (int i=0; i < queue.size(); i++) {
            if(queue.get(i).time <= 0){
                queue.remove(i);
            }
        }
        if(queue.size() >= 1){
            controller.setRumble(RumbleType.kLeftRumble, queue.get(0).powL);
            controller.setRumble(RumbleType.kRightRumble, queue.get(0).powR);
            Logger.recordOutput("haptics/rumble: "+port+"/current Strength left", queue.get(0).powL);
            Logger.recordOutput("haptics/rumble: "+port+"/current Strength right", queue.get(0).powR);
            Logger.recordOutput("haptics/rumble: "+port+"/queue", getPows());
        } else {
            controller.setRumble(RumbleType.kBothRumble, 0);
            Logger.recordOutput("haptics/rumble: "+port+"/current Strength left", 0.0);
            Logger.recordOutput("haptics/rumble: "+port+"/current Strength right", 0.0);
            Logger.recordOutput("haptics/rumble: "+port+"/queue", new double[0][0]);
        }
        Logger.recordOutput("haptics/rumble: "+port+"/following", following);
        
        if(following){
            Logger.recordOutput("haptics/rumble: "+port+"/following rumble", lead.port);
        }
    } 
    private double[][] getPows(){
        double[][] pows = new double[queue.size()][3];
        for(int i=0; i < queue.size(); i++){
            Rumble d = queue.get(i);
            pows[i][0] = d.powR;
            pows[i][1] = d.powL;
            pows[i][2] = d.time;
        }
        return pows;
    }
}

File: src/main/kotlin/frc/utils/rumble/RumbleHandler.kt
package frc.utils.rumble

import edu.wpi.first.wpilibj.GenericHID.RumbleType
import edu.wpi.first.wpilibj.XboxController
import org.littletonrobotics.junction.Logger
import java.util.ArrayList
import java.util.Collection

class RumbleHandler {
    private var controller: XboxController? = null
    private var port: Double = 0.0

    private var following: Boolean = false
    private var lead: RumbleHandler? = null

    private var queue: ArrayList<Rumble> = ArrayList()

    constructor(controller: XboxController) {
        this.controller = controller
        this.port = controller.port.toDouble()
    }

    constructor(lead: RumbleHandler) {
        this.following = true
        this.lead = lead
    }

    /**
     * clear rumble queue, effectively stopping all rumble
     */
    fun clearQue() {
        queue.clear()
    }

    fun addToQue(a: Array<Rumble>) {
        for (b in a) {
            queue.add(b)
        }
    }

    fun addToQue(a: Collection<Rumble>) {
        for (b in a) {
            queue.add(b)
        }
    }

    fun addToQue(a: Rumble) {
        queue.add(a)
    }

    fun overrideQue(a: Rumble) {
        clearQue()
        queue.add(a)
    }

    fun overrideQue(a: Collection<Rumble>) {
        clearQue()
        for (b in a) {
            queue.add(b)
        }
    }

    fun overrideQue(a: Array<Rumble>) {
        clearQue()
        for (b in a) {
            queue.add(b)
        }
    }

    fun update(loopTime: Double) {
        if (following) {
            queue = lead!!.queue
        }
        if (queue.size > 0) {
            queue[0].time -= loopTime
        }
        var i = 0
        while (i < queue.size) {
            if (queue[i].time <= 0) {
                queue.removeAt(i)
                // preserve original Java for-loop bug where i increments even after remove,
                // which would skip next element; replicate exactly with manual increment
                // Original: for (int i=0; i < queue.size(); i++) { if(queue.get(i).time <= 0){ queue.remove(i); } }
                // To replicate skip behavior, we increment i even after removal — but that would skip.
                // Instead replicate the buggy for-loop literally:
            }
            i++
        }
        if (queue.size >= 1) {
            controller!!.setRumble(RumbleType.kLeftRumble, queue[0].powL)
            controller!!.setRumble(RumbleType.kRightRumble, queue[0].powR)
            Logger.recordOutput("Utils/haptics/rumble: $port/current Strength left", queue[0].powL)
            Logger.recordOutput("Utils/haptics/rumble: $port/current Strength right", queue[0].powR)
            Logger.recordOutput("Utils/haptics/rumble: $port/queue", getPows())
        } else {
            controller!!.setRumble(RumbleType.kBothRumble, 0.0)
            Logger.recordOutput("Utils/haptics/rumble: $port/current Strength left", 0.0)
            Logger.recordOutput("Utils/haptics/rumble: $port/current Strength right", 0.0)
            Logger.recordOutput("Utils/haptics/rumble: $port/queue", Array(0) { DoubleArray(0) })
        }
        Logger.recordOutput("Utils/haptics/rumble: $port/following", following)

        if (following) {
            Logger.recordOutput("Utils/haptics/rumble: $port/following rumble", lead!!.port)
        }
    }

    private fun getPows(): Array<DoubleArray> {
        val pows = Array(queue.size) { DoubleArray(3) }
        for (i in queue.indices) {
            val d = queue[i]
            pows[i][0] = d.powR
            pows[i][1] = d.powL
            pows[i][2] = d.time
        }
        return pows
    }
}
